compiler/hsSyn/HsTypes.hs

```diff
@@ -403,13 +404,15 @@ instance OutputableBndr HsIPName where
 data HsTyVarBndr pass
   = UserTyVar        -- no explicit kinding
          (Located (IdP pass))
+         ArgFlag
```

Comment:
I think this is a good idea, and I had planned on doing something similar myself. I'd thought to do it like:

```
data InvisibleFlag = Specified | Inferred

data ArgFlag = Required | Invisible InvisibleFlag
```

But I like the GADT approach better; it probably requires less refactoring, and the types say more clearly what's going on.

So it's okay to use GADTs inside GHC? I'm not clear on what extensions are considered okay and which are frowned upon, or complicate bootstrapping. I'm guessing TH is a no, but other than that?

Reply:
GHC has a 2-release bootstrap policy. Having released 8.2, we can now use features that exist in 8.0. So, technically, you could use TypeInType. But don't. :) On the other hand, TypeApplications would be reasonable. GADTs are very very reasonable; this wouldn't be the first GADT.

I also thought of your solution, but it is less performant, because GHC has to pointer-chase to discover which invisible ArgFlag you have.